src/Orleans.Transactions/State/TransactionalStateStorageProviderWrapper.cs

```diff
@@ -0,0 +1,83 @@
+            this.stateStorages = new ConcurrentDictionary<string, IStorage<TransactionalStateRecord<TState>>>();
+        }
+
+        public async Task<TransactionalStorageLoadResponse<TState>> Load(string stateName)
+        {
+            IStorage<TransactionalStateRecord<TState>> stateStorage = GetStateStorage(stateName);
+            await stateStorage.ReadStateAsync();
+            return new TransactionalStorageLoadResponse<TState>(stateStorage.Etag, stateStorage.State.CommittedState, stateStorage.State.Metadata, stateStorage.State.PendingStates);
+        }
+
+        public async Task<string> Persist(string stateName, string expectedETag, string metadata, List<PendingTransactionState<TState>> statesToPrepare)
+        {
+            IStorage<TransactionalStateRecord<TState>> stateStorage = GetStateStorage(stateName);
+            if (stateStorage.Etag != expectedETag)
+                throw new ArgumentException(nameof(expectedETag), "Etag does not match");
+            stateStorage.State.Metadata = metadata;
+            foreach(PendingTransactionState<TState> pendingState in statesToPrepare.Where(s => !stateStorage.State.PendingStates.Contains(s)))
+            {
+                stateStorage.State.PendingStates.Add(pendingState);
+            }
+            await stateStorage.WriteStateAsync();
+            return stateStorage.Etag;
+        }
+
+        public async Task<string> Confirm(string stateName, string expectedETag, string metadata, string transactionIdToCommit)
+        {
+            IStorage<TransactionalStateRecord<TState>> stateStorage = GetStateStorage(stateName);
+            if (stateStorage.Etag != expectedETag)
+                throw new ArgumentException(nameof(expectedETag), "Etag does not match");
+            stateStorage.State.Metadata = metadata;
+            PendingTransactionState<TState> committedState = stateStorage.State.PendingStates.FirstOrDefault(pending => transactionIdToCommit == pending.TransactionId);
+            if (committedState != null)
+            {
+                stateStorage.State.CommittedTransactionId = committedState.TransactionId;
+                stateStorage.State.CommittedState = committedState.State;
+                stateStorage.State.PendingStates = stateStorage.State.PendingStates.Where(pending => pending.SequenceId > committedState.SequenceId).ToList();
+            }
+            await stateStorage.WriteStateAsync();
+            return stateStorage.Etag;
+        }
+
+        private IStorage<TransactionalStateRecord<TState>> GetStateStorage(string stateName)
+        {
+            return this.stateStorages.GetOrAdd(stateName, name => new StateStorageBridge<TransactionalStateRecord<TState>>(name, this.context.GrainInstance.GrainReference, storageProvider));
+        }
+    }
+
+    [Serializable]
+    public class TransactionalStateRecord<TState>
+        where TState : class, new()
+    {
+        public TState CommittedState { get; set; } = new TState();
```

Comment:
> Be called every time we initialize this object?
This field 'should' only be initialized once on initial create, but the behavior is provider dependent.

> Most of the times will be overwritten, isn't it?
It will be overridden on any commit, and possibly on load, depending on provider implementation.

There is likely some unnecessary allocations due to this pattern, but state management with providers has been a source of bugs before and this is the pattern we've used to address this in the past (see PubSubGrainState and it's history).  Since this class is specific to how this feature wraps the existing storage providers, it's safest to follow the proven patterns.  If we formalize around a more efficient state management for storage providers, and retrofit all of the providers to follow that pattern, we can address it here at that time.